checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java

```diff
@@ -1,176 +1,850 @@
-      successorBlock.add(ccur.getElseSuccessor());
+      if (nodeAndElementResourceAliased(obligations, arg, paramElt)) {
+        List<String> mustCallValues = resourceLeakAtf.getMustCallValues(paramElt);
+        // TODO: generalize this method to MustCall annotations with more than one element.
+        assert mustCallValues.size() <= 1 : "TODO: Handle larger must-call values sets";
+        Set<ResourceAlias> nodeAliases = getResourceAliasOfNode(obligations, arg);
+        for (ResourceAlias resourceAlias : nodeAliases) {
+          AnnotationMirror cmAnno = getCalledMethodsAnno(invocation, resourceAlias.reference);
+          if (mcca.calledMethodsSatisfyMustCall(mustCallValues, cmAnno)) {
+            addOwningToParam(i + 1);
+            break outerLoop;
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Returns the set of resource aliases associated with the given node, by looking up the
+   * corresponding obligation in the given set of obligations.
+   *
+   * @param obligations the set of obligations to search in
+   * @param node the node whose resource aliases are to be returned
+   * @return the resource aliases associated with the given node, or an empty set if the node has
+   *     none
+   */
+  private Set<ResourceAlias> getResourceAliasOfNode(Set<Obligation> obligations, Node node) {
+    Node tempVar = mcca.getTempVarOrNode(node);
+    if (!(tempVar instanceof LocalVariableNode)) {
+      return Collections.emptySet();
+    }
 
-    } else {
-      if (!(cur instanceof SingleSuccessorBlock)) {
-        throw new BugInCF("BlockImpl is neither a conditional block nor a SingleSuccessorBlock");
+    Obligation argumentObligation =
+        MustCallConsistencyAnalyzer.getObligationForVar(obligations, (LocalVariableNode) tempVar);
+    if (argumentObligation == null) {
+      return Collections.emptySet();
+    }
+    return argumentObligation.resourceAliases;
+  }
+
+  /**
+   * Computes @Owning annotations for the enclosing formal parameter or fields:
+   *
+   * <ul>
+   *   <li>If a formal parameter is passed as an owning parameter, it adds the @Owning annotation to
+   *       that formal parameter (see {@link #inferOwningParamsViaOwnershipTransfer}).
+   *   <li>It calls {@link #computeOwningForReceiver} to verify if the receiver of the method
+   *       represented by {@code invocation} qualifies as a candidate owning field, and if the
+   *       method invocation satisfies the field's must-call obligation. If these conditions are
+   *       met, the field is added to the {@link #disposedFields} set.
+   *   <li>It calls {@link #computeOwningForArgsOfCall} to inspect the method represented by the
+   *       given MethodInvocationNode for any indirect calls within it. The method analyzes the
+   *       called-methods set of each argument after the call and computes the @Owning annotation to
+   *       the field or parameter passed as an argument to this call.
```

Comment:
I can't understand what this means.  What is an "indirect call"?

Reply:
It points to the calls on the argument inside the callee method by checking the called-methods set. The comment contains implementation details, so I revise this doc.